gapis/api/vulkan/state_rebuilder.go

```diff
@@ -1497,80 +1461,157 @@ func (sb *stateBuilder) createImage(img *ImageObject) {
 	}
 	// We have to handle the above cases at some point.
 
-	offset := VkDeviceSize(0)
-	{
-		for _, rng := range opaqueRanges {
-			for i := uint32(0); i < rng.LevelCount; i++ {
-				mipLevel := rng.BaseMipLevel + i
-				e := sb.levelSize(img.Info.Extent, img.Info.Format, mipLevel)
-				for j := uint32(0); j < rng.LayerCount; j++ {
-					layer := rng.BaseArrayLayer + j
-					copies = append(copies, VkBufferImageCopy{
-						offset,
-						0,
-						0,
-						VkImageSubresourceLayers{
-							img.ImageAspect,
-							mipLevel,
-							layer,
-							1,
-						},
-						VkOffset3D{
-							0, 0, 0,
-						},
-						VkExtent3D{
-							uint32(e.width),
-							uint32(e.height),
-							uint32(e.depth),
-						},
-					})
-					data := img.Layers.Get(layer).Levels.Get(mipLevel).Data.MustRead(sb.ctx, nil, sb.oldState, nil)
-					contents = append(contents, data...)
-					// len(data) must be equal to e.alignedLevelSize
-					offset += VkDeviceSize(e.alignedLevelSize)
-				}
-			}
+	// Prime the image data according to the usage bits
+	transDstBit := VkImageUsageFlags(VkImageUsageFlagBits_VK_IMAGE_USAGE_TRANSFER_DST_BIT)
+	attBits := VkImageUsageFlags(VkImageUsageFlagBits_VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VkImageUsageFlagBits_VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT)
+	storageBit := VkImageUsageFlags(VkImageUsageFlagBits_VK_IMAGE_USAGE_STORAGE_BIT)
+
+	primeByBufCopy := img.Info.Usage&transDstBit != VkImageUsageFlags(0)
+	primeByRendering := (!primeByBufCopy) && (img.Info.Usage&attBits != VkImageUsageFlags(0))
+	primeByShaderCopy := (!primeByBufCopy) && (!primeByRendering) && (img.Info.Usage&storageBit != VkImageUsageFlags(0))
+
+	copyDstImgs := []*ImageObject{}
+	copies := map[*ImageObject][]VkBufferImageCopy{}
+
+	if primeByBufCopy {
+		// Copy to the state block image directly
+		copyDstImgs = []*ImageObject{img}
+	} else if primeByRendering || primeByShaderCopy {
+		// Copy to a staging image. Later the staging image will be used to
+		// render to the state block image, or copied into the state block image.
+		stagingImgs, err := helper.allocateStagingImagesFor(img)
+		if err != nil {
+			log.E(sb.ctx, "[Creating staging image for %v] %v", img.VulkanHandle, err)
+			return
 		}
+		if len(stagingImgs) == 0 {
+			log.E(sb.ctx, "No staging image created")
+			return
+		}
+		copyDstImgs = stagingImgs
+
+	} else {
+		// There is no way to prime the data
```

Comment:
Maybe we want to issue a warning. (Data in image unprimed, since there was no valid way to add data in the first place)